### PR TITLE
TRUNK-6047: Add utility methods to invalidate database caches

### PR DESCRIPTION
**Pull request by docjon09**
## Description of what I changed

* Update ContextTest.java

* Update Context.java

## Issue I worked on

see https://issues.openmrs.org/browse/TRUNK-6047

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by docjon09** [comment]
The above changes compile and the tests I have written for them are passed. However, the ModuleFactoryTest class is showing error. As you can see, I haven't changed that class so I have no idea why this is happening.

**Comment by tendomart** [comment]
There is a problem at   /home/runner/work/openmrs-core/openmrs-core/api/src/main/java/org/openmrs/api/context/Context.java:[1441,25] unreachable statement

please run `mvn clean package` after making changes and before doing a commit to avoid these failures.

File: api/src/main/java/org/openmrs/api/context/Context.java [hunk 1]
```diff
@@ -1427,4 +1428,49 @@ public static void setUseSystemClassLoader(boolean useSystemClassLoader) {
...
+			return;
+		}
+		log.debug("Clearing DB cache for entities of type: {}", entityClass);
+		sessionFactory.getCache().evictEntityRegion(entityClass);
+		sessionFactory.getCache().evictCollectionRegions();
+		sessionFactory.getCache().evictQueryRegions();
+	}
+
+	/**
+	 * Evict data from all cache regions.
```
[the remark this answers is not in the document]
Comment (docjon09) [reply api/src/main/java/org/openmrs/api/context/Context.java]: done. thanks

File: api/src/test/java/org/openmrs/api/context/ContextTest.java [hunk 2]
```diff
@@ -238,4 +253,92 @@ public void becomeUser_shouldChangeLocaleWhenBecomeAnotherUser() {
...
 	}
+	
+	/**
+	 * @see Context#evictSingleEntity(SessionFactory, Class, String)
+	 */
+	@Test
+	public void evictSingleEntity_shouldClearSingleEntityFromCaches(){
+		
+		PersonName name = Context.getPersonService().getPersonName(PERSON_NAME_ID_2);
+		if(name == null){
```
[the remark this answers is not in the document]
Comment (docjon09) [reply api/src/test/java/org/openmrs/api/context/ContextTest.java]: done. thanks

File: api/src/test/java/org/openmrs/api/context/ContextTest.java [hunk 3]
```diff
@@ -238,4 +253,92 @@ public void becomeUser_shouldChangeLocaleWhenBecomeAnotherUser() {
...
+	}
+
+	/**
+	 * @see Context#evictAllEntities(SessionFactory, Class)
+	 */
+	@Test
+	public void evictAllEntities_shouldClearAllEntityFromCaches(){
+		PersonName name1 = Context.getPersonService().getPersonName(PERSON_NAME_ID_2);
+		PersonName name2 = Context.getPersonService().getPersonName(PERSON_NAME_ID_9349);
+		//Load the person so that the names are also stored  in person names collection region
```
[the remark this answers is not in the document]
Comment (docjon09) [reply api/src/test/java/org/openmrs/api/context/ContextTest.java:293]: done. thanks

**Comment by docjon09** [comment]
> There is a problem at **/home/runner/work/openmrs-core/openmrs-core/api/src/main/java/org/openmrs/api/context/Context.java:[1441,25] unreachable statement**
> 
> please run `mvn clean package` after making changes and before doing a commit to avoid these failures.

I did run the test locally. But I accidentally pushed the wrong branch. I have updated it now and it passes all the tests. 
I thought I knew how to work with github but I still have to learn a lot.

File: api/src/main/java/org/openmrs/api/context/Context.java [hunk 4]
```diff
@@ -1427,4 +1428,47 @@ public static void setUseSystemClassLoader(boolean useSystemClassLoader) {
...
 	}
+	
+	/**
+	 * Evicts the entity data for a particular entity "instance".
+	 *
+	 * @param sessionFactory
+	 * @since 2.6.0
+	 */
+	public static void evictSingleEntity(SessionFactory sessionFactory, Class<?> entityClass, Integer id) {
+		if((id == null) || (entityClass == null)) {
```
Comment (tendomart) [inline api/src/main/java/org/openmrs/api/context/Context.java:1439]: `if((StringUtils.isNotEmpty(id)) || (StringUtils.isNotEmpty(entityClass))) `

File: api/src/main/java/org/openmrs/api/context/Context.java [hunk 5]
```diff
@@ -1427,4 +1428,47 @@ public static void setUseSystemClassLoader(boolean useSystemClassLoader) {
...
+	}
+
+	/**
+	 * Evicts all entity data from the given region (i.e. for all entities of type).
+	 *
+	 * @param sessionFactory
+	 * @since 2.6.0
+	 */
+	public static void evictAllEntities(SessionFactory sessionFactory, Class<?> entityClass) {
+		if(entityClass == null){
```
Comment (tendomart) [inline api/src/main/java/org/openmrs/api/context/Context.java:1455]: change to this `if(StringUtils.isNotEmpty(entityClass)){}`

as well

generally avoid if statements which resolve to null

**Comment by tendomart** [comment]
And pay attention to all the comments which  I made on this PR.

File: api/src/main/java/org/openmrs/api/context/Context.java (hunk 5 above)
Comment (docjon09) [reply api/src/main/java/org/openmrs/api/context/Context.java:1455]: I guess that method is only applicable for strings. When I changed it the compilation failed.

**Pull request closed without merging**